src/overview.md

```diff
@@ -0,0 +1,322 @@
+    - Lexer definition: [`librustc_lexer`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lexer/index.html)
+    - Main entry point: **TODO**
+- Parsing: Parse the stream of tokens to an Abstract Syntax Tree (AST)
+    - Guide: [Lexing and Parsing](https://rust-lang.github.io/rustc-guide/the-parser.html)
+    - Parser definition: [`librustc_parse`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/index.html)
+    - Main entry point: **TODO**
+    - AST definition: [`librustc_ast`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/index.html)
+- The High Level Intermediate Representation (HIR)
+    - Guide: [The HIR](https://rust-lang.github.io/rustc-guide/hir.html)
+    - Guide: [Identifiers in the HIR](https://rust-lang.github.io/rustc-guide/hir.html#identifiers-in-the-hir)
+    - Guide: [The HIR Map](https://rust-lang.github.io/rustc-guide/hir.html#the-hir-map)
+    - Guide: [Lowering AST to HIR](https://rust-lang.github.io/rustc-guide/lowering.html)
+    - How to view HIR representation for your code `cargo rustc -- -Zunpretty=hir-tree`
+    - Rustc HIR definition: [`rustc_hir`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/index.html)
+    - Main entry point: **TODO**
+- Type Inference
+    - Guide: [Type Inference](https://rust-lang.github.io/rustc-guide/type-inference.html)
+    - Guide: [The ty Module: Representing Types](https://rust-lang.github.io/rustc-guide/ty.html) (semantics)
+    - Main entry point: **TODO**
+- The Mid Level Intermediate Representation (MIR)
+    - Guide: [The MIR (Mid level IR)](https://rust-lang.github.io/rustc-guide/mir/index.html)
+    - Definition: [`librustc/mir`](https://github.com/rust-lang/rust/tree/master/src/librustc/mir)
+    - Definition of source that manipulates the MIR: [`librustc_mir`](https://github.com/rust-lang/rust/tree/master/src/librustc_mir)
+    - Main entry point: **TODO**
+- The Borrow Checker
+    - Guide: [MIR Borrow Check](https://rust-lang.github.io/rustc-guide/borrow_check.html)
+    - Definition: [`rustc_mir/borrow_check`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/borrow_check/index.html)
+    - Main entry point: **TODO**
+- MIR Optimizations
+    - Guide: [MIR Optimizations](https://rust-lang.github.io/rustc-guide/mir/optimizations.html)
+    - Definition: [`rustc_mir/transform`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/transform/index.html) **TODO: is this correct?**
+    - Main entry point: **TODO**
+- Code Generation
+    - Guide: [Code Generation](https://rust-lang.github.io/rustc-guide/codegen.html)
+    - Guide: [Generating LLVM IR](https://rust-lang.github.io/rustc-guide/codegen.html#generating-llvm-ir) - **TODO: this is not available yet**
+    - Generating Machine Code from LLVM IR with LLVM - **TODO: reference?**
+    - Main entry point MIR -> LLVM IR: **TODO**
+    - Main entry point LLVM IR -> Machine Code **TODO**
```

Comment:
Is it [`rustc_codegen_ssa::base::codegen_crate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/base/fn.codegen_crate.html)?
```suggestion
    - Main entry point LLVM IR -> Machine Code: [`rustc_codegen_ssa::base::codegen_crate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/base/fn.codegen_crate.html)
```

Reply:
@bjorn3 can you confirm?

Reply:
Also, as I understand it, starting codegen also transitively causes monomorphization for the CGU and then generates LLVM IR from the collector's results.

Reply:
> @bjorn3 can you confirm?

`rustc_codegen_ssa::base::codegen_crate` is responsible for calling the right functions to translate everything in a codegen unit to LLVM ir. It then starts a background thread for that codegen unit to optimize it and write the resulting object file. After all threads are started it returns. You then have to call [`rustc_codegen_ssa::back::write::OngoingCodegen::join`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/back/write/struct.OngoingCodegen.html#method.join) from [`rustc_codegen_ssa::traits::CodegenBackend::join_codegen](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/traits/trait.CodegenBackend.html#tymethod.join_codegen) to join all threads. That function then returns all compiled modules. This split exists to be able to free the `TyCtxt` before all codegen units are optimized and written, potentially reducing max RSS.

> Also, as I understand it, starting codegen also transitively causes monomorphization for the CGU and then generates LLVM IR from the collector's results.

Monomorphization happens lazily. Every time something part of the MIR (eg a statement, or a part of a statement) is codegened, `FunctionCx::monomorphize` is called on that part.